### ISSUE TITLE
Some SocketOptions not supported on Windows

### ISSUE THREAD

**Issue by joakime** [eclipse/jetty.project#6661]
**Jetty version(s)**
10.0.x

**Java version/vendor** `(use: java -version)`
11.0.12

**OS type/version**
Windows 10

**Description**

The recent work on Issue #6372 is not compatible with Windows 10.

```
Caused by: java.lang.UnsupportedOperationException: 'SO_REUSEPORT' not supported
	at java.base/sun.nio.ch.ServerSocketChannelImpl.setOption(ServerSocketChannelImpl.java:151)
	at org.eclipse.jetty.server.ServerConnector.openAcceptChannel(ServerConnector.java:338)
```

**Comment by joakime** [comment]
Opened PR #6663

**Comment by delanym** [comment]
Did this go in? I'm still getting this on Jetty 11.0.12 
```
java.lang.UnsupportedOperationException: 'SO_REUSEPORT' not supported
    at java.base/sun.nio.ch.ServerSocketChannelImpl.setOption(ServerSocketChannelImpl.java:151)
    at org.eclipse.jetty.server.ServerConnector.setSocketOption(ServerConnector.java:355)
```

**Comment by sbordet** [comment]
The exception is logged at DEBUG level since 11.0.7.
If you have DEBUG logs enabled, you will "get" it.
Otherwise, you have to be more specific about where you "get" it.

**Comment by joakime** [comment]
@delanym That is Jetty reporting (via debug level logging event) that your JVM is unable to support SO_REUSEPORT on that Socket type.

Here's the code in 11.0.12 ...

https://github.com/eclipse/jetty.project/blob/jetty-11.0.12/jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java#L351-L362

And it being used to attempt to set SO_REUSEPORT

https://github.com/eclipse/jetty.project/blob/jetty-11.0.12/jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java#L335

**Comment by delanym** [comment]
Thanks I'll see whats going on with the JDK. From my miserable understanding of Java serving up a stack exception is a costly affair so i'd like to avoid that, and reuse the port if possible.

**Comment by joakime** [comment]
> From my miserable understanding of Java serving up a stack exception is a costly affair so i'd like to avoid that

This happens on the `ServerConnector` once only when the Listening socket is setup to accept incoming socket connections.
If you were using something like Java 1.4, it could be costly.  But that's not been true for a long time.  especially so with Java 11 (the oldest version you could be using).

> and reuse the port if possible.

Windows has all sorts of wonky and hard to follow rules on what socket options can be used under what conditions.
You'll probably find that there's nothing you really can do on the Java or Jetty side, as ultimately Java is just calling the OS level APIs to set the Socket options and are being rejected by those same APIs.

If you take look at C# or Dot-Net you'll find that at least those runtimes will also fail on setting SO_REUSEADDR and say something along the lines of "System.PlatformNotSupportedException: Socket options not supported on this platform, as they are invalid for use after a failed connection attempt"

**Comment by joakime** [comment]
@delanym looks like many projects (even non-Java projects) do not set `SO_REUSEADDR` or `SO_EXCLUSIVEADDRUSE` (or `SO_REUSEPORT`) on Windows.

* https://github.com/python-trio/trio/issues/928
* https://github.com/tornadoweb/tornado/issues/550
* https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ/14388707#14388707
* https://github.com/python-trio/trio/pull/1108
* https://github.com/rpclib/rpclib/pull/278
* https://github.com/search?q=SO_REUSEADDR+windows+in%3Atitle+is%3Apr&type=Issues

We attempt to set it, and if unable we log it, and continue running.
